src/Simulation/qdk_sim_rs/src/tableau.rs

```diff
@@ -0,0 +1,298 @@
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "Stabilizer tableau on {} qubits:\n{:?}",
+            self.n_qubits, self.table
+        )
+    }
+}
+
+impl Tableau {
+    const fn idx_phase(&self) -> usize {
+        2 * self.n_qubits
+    }
+
+    /// Returns a new stabilizer tableau representing
+    /// the $\ket{00\cdots 0}$ state of an $n$-qubit
+    /// register.
+    pub fn new(n_qubits: usize) -> Self {
+        Tableau {
+            n_qubits,
+            table: Array::from_shape_fn((2 * n_qubits, 2 * n_qubits + 1), |(i, j)| i == j),
+        }
+    }
+
+    fn idx_x(&self, idx_col: usize) -> usize {
+        idx_col
+    }
+
+    fn idx_z(&self, idx_col: usize) -> usize {
+        idx_col + self.n_qubits
+    }
+
+    /// Returns the determinstic result that would be obtained from measuring
+    /// the given qubit in the 𝑍 basis, or `None` if the result is random.
+    fn determinstic_result(&self, idx_target: usize) -> Option<bool> {
+        let determined = !self
+            .table
+            .slice(s![self.n_qubits.., idx_target])
+            .iter()
+            .any(|b| *b);
+        if determined {
+            // What was it determined to be?
+            let mut vector: Array1<bool> = Array::default(2 * self.n_qubits + 1);
+            for idx_destabilizer in 0..self.n_qubits {
+                if self.table[(idx_destabilizer, self.idx_x(idx_target))] {
+                    set_vec_to_row_sum(&mut vector, &self.table, idx_destabilizer + self.n_qubits);
+                }
+            }
+            Some(vector[2 * self.n_qubits])
+        } else {
+            None
+        }
+    }
+}
+
+impl Tableau {
+    /// Asserts whether a hypothetical single-qubit $Z$-basis measurement
+    /// would agree with an expected result.
+    ///
+    /// If the assertion would pass, `Ok(())` is returned, otherwise an [`Err`]
+    /// describing the assertion failure is returned.
+    pub fn assert_meas(&self, idx_target: usize, expected: bool) -> Result<(), String> {
+        let actual = self.determinstic_result(idx_target).ok_or(format!(
+            "Expected {}, but measurement result would be random.",
+            expected
+        ))?;
+        if actual != expected {
+            Err(format!(
+                "Expected {}, but measurement result would actually be {}.",
+                expected, actual
+            ))
+        } else {
+            Ok(())
+        }
+    }
+}
+
+#[cfg_attr(feature = "python", pymethods)]
+impl Tableau {
+    /// Returns a serialization of this stabilizer tableau as a JSON object.
+    #[cfg(feature = "python")]
+    pub fn as_json(&self) -> PyResult<String> {
+        serde_json::to_string(self)
+            .map_err(|e| PyErr::from(Error::new(ErrorKind::Other, e.to_string())))
+    }
+
+    /// Applies a Hadamard operation in-place to the given qubit.
+    pub fn apply_h_mut(&mut self, idx_target: usize) {
```

Comment:
nitpick: is there any reason besides establishing a convention that these methods can't just avoid the `_mut` in their names? I was a little surprised that there isn't an `apply_h` to match `apply_h_mut`. Do you expect immutable variants of these to exist in the future (possibly by returning the updated table instead of modifying in-place)?

Reply:
I chose the names here to parallel usage in `ndarray`, where `_mut` is often used to indicate in-place even when no immutable variant exists. That said, I think it would be worthwhile at some point to expand to include `apply_h` as well as `apply_h_mut`.